services/sharedchannel/sync_recv.go

```diff
@@ -28,223 +28,211 @@ func (scs *Service) onReceiveSyncMessage(msg model.RemoteClusterMsg, rc *model.R
+		scs.server.GetLogger().Log(mlog.LvlSharedChannelServiceDebug, "Sync msg received",
+			mlog.String("post_id", sm.PostId),
+			mlog.String("channel_id", sm.ChannelId),
+			mlog.Int("reaction_count", len(sm.Reactions)),
+			mlog.Int("user_count", len(sm.Users)),
+			mlog.Bool("has_post", sm.Post != nil),
+		)
 
 		if channel == nil {
-			if channel, err = scs.server.GetStore().Channel().Get(sm.Post.ChannelId, true); err != nil {
+			if channel, err = scs.server.GetStore().Channel().Get(sm.ChannelId, true); err != nil {
 				// if the channel doesn't exist then none of these sync messages are going to work.
-				return postErrors, 0, fmt.Errorf("channel not found processing sync messages: %w", err)
+				return fmt.Errorf("channel not found processing sync messages: %w", err)
+			}
+		}
+
+		if sm.Post != nil {
+			if _, ok := chanToTeamMap[sm.Post.ChannelId]; !ok {
+				team, err2 := scs.server.GetStore().Channel().GetTeamForChannel(sm.Post.ChannelId)
+				if err2 != nil {
+					scs.server.GetLogger().Log(mlog.LvlSharedChannelServiceError, "Error getting Team for Channel",
+						mlog.String("ChannelId", sm.Post.ChannelId),
+						mlog.String("PostId", sm.Post.Id),
+						mlog.Err(err2),
+					)
+					postErrors = append(postErrors, sm.Post.Id)
+					continue
+				}
+				chanToTeamMap[sm.Post.ChannelId] = team
+			}
+			sm.Post.Message = scs.processPermalinkFromRemote(sm.Post, chanToTeamMap[sm.Post.ChannelId])
+		}
+
+		// add/update users first
+		for _, user := range sm.Users {
+			if userSaved, err := scs.upsertSyncUser(user, rc); err != nil {
+				scs.server.GetLogger().Log(mlog.LvlSharedChannelServiceError, "Error upserting sync user",
+					mlog.String("post_id", sm.PostId),
+					mlog.String("channel_id", sm.ChannelId),
+					mlog.String("user_id", user.Id),
+					mlog.Err(err))
+			} else {
+				usersSyncd = append(usersSyncd, userSaved.Id)
+				scs.server.GetLogger().Log(mlog.LvlSharedChannelServiceDebug, "User upserted via sync",
+					mlog.String("channel_id", sm.ChannelId),
+					mlog.String("user_id", user.Id))
 			}
 		}
 
-		if err := scs.server.GetStore().SharedChannel().UpsertPost(sm.Post); err != nil {
-			scs.server.GetLogger().Log(mlog.LvlSharedChannelServiceError, "Error saving sync Post",
-				mlog.String("remote", rc.DisplayName),
-				mlog.String("ChannelId", sm.Post.ChannelId),
-				mlog.String("PostId", sm.Post.Id),
-				mlog.Err(err),
-			)
-			postErrors = append(postErrors, sm.Post.Id)
-			continue
+		// add post (may be nil if only reactions changed)
+		if sm.Post != nil {
+			rpost, err := scs.upsertSyncPost(sm.Post, channel, rc)
+			if err != nil {
+				postErrors = append(postErrors, sm.Post.Id)
+				scs.server.GetLogger().Log(mlog.LvlSharedChannelServiceError, "Error upserting sync post",
+					mlog.String("post_id", sm.Post.Id),
+					mlog.String("channel_id", sm.Post.ChannelId),
+					mlog.Err(err))
+			} else if lastSyncAt < rpost.UpdateAt {
+				lastSyncAt = rpost.UpdateAt
+			}
```

Comment:
So if somehow the `upsertSyncPost` fails and a subsequent one succeeds, could setting `lastSyncAt` cause losing posts? (same argument for reactions below)

Reply:
Yes it can.  I thought a lot about that, and sometime in the future I will need a retry on sync'ing posts/reactions with a max count for retries.  On the receiving side, if we don't move on by setting lasySyncAt and the error repeats we will be stuck at that point.  We do currently send the failed post ids back to the sender, so the retry can be added there.  I don't think we want to block it here.